### DISCUSSION TITLE
How to load data from SQL and a pure Ruby source in a way that allows proper testing?

### DISCUSSION THREAD

**Discussion by 3ynm** [jeremyevans/sequel#2063] [General]
I need to create a model with 2 data sources: SQL and a ruby array.

I've been able to do what I need using the `after_initialize` plugin, but testing its behavior has been challenging, so, I'm looking for suggestions on alternative implementations.

Here is what I'm doing:

```ruby
# DB fields: id, stock
class Product < Sequel::Model
  plugin :after_initialize
  attr_reader :name, :description, :price

  def after_initialize
    return unless @values[:id]

    Product.static_data.find{ |p| p[:id] == @values[:id] }.each do |name, value|
      next if name == :id

      instance_variable_set("@#{name}", value)
    end
  end

  def self.static_data
    require_relative('../data/products')
    PRODUCTS
  end
end
```

An example `../data/product.rb` would be:

```ruby
PRODUCTS = [
  {
    id: 'test_product',
    name: 'Test product',
    description: 'Test product description',
    price: 100_000
  } #, { ... }
]
```

For testing, I could replace the `Product.static_data` method on my tests. But I'm not sure that's a good strategy (would change the source data for all tests).

Other approach, instead of using after_initialize, could be defining a `load` method to be executed after the loading of data from database. I think that is sub-optimal since I will always load data from both sources.

```ruby
class Product < Sequel::Model
  # ...
  def load(static_data = Product.static_data)
    return self unless @values[:id]

    static_data.find{ |p| p[:id] == @values[:id] }.each do |name, value|
      next if name == :id

      instance_variable_set("@#{name}", value)
    end
    self
  end
  # ...
end
```

What would be your approach to the matter?

**Comment by JonathanFrias** [comment]
I would load the static data into a table, and specify an eager-loadable association. That way you let your 2 data sources be 2 data sources, rather than something in between. 

Just FYI, doing a `static_data.find` is a full scan that will not scale well, in addition to keeping potentially unbounded amounts of data in memory.

**Comment by 3ynm** [comment]
Turns out I was stubbing things wrong (or at least not as minitest expects me to do so). So, I can stub now and I don't need to change my code for testing.

However, I see the problem of having 2 data sources. I'll probably refactor on a future version. I'm testing previous to refactoring actually.
